std/algorithm/sorting.d

```diff
@@ -1260,40 +1260,154 @@ private size_t getPivot(alias less, Range)(Range r)
     return mid;
 }
 
-private void optimisticInsertionSort(alias less, Range)(Range r)
+/*
+Sorting routine that is optimized for short ranges. Note: uses insertion sort
+going downward. Benchmarked a similar routine that goes upward, for some reason
+it's slower.
+*/
+private void shortSort(alias less, Range)(Range r)
 {
     import std.algorithm.mutation : swapAt;
-
     alias pred = binaryFun!(less);
-    if (r.length < 2)
+
+    switch (r.length)
     {
-        return;
+        case 0: case 1:
+            return;
+        case 2:
+            if (pred(r[1], r[0])) r.swapAt(0, 1);
+            return;
+        case 3:
+            if (pred(r[2], r[0]))
+            {
+                if (pred(r[0], r[1]))
+                {
+                    r.swapAt(0, 1);
+                    r.swapAt(0, 2);
+                }
+                else
+                {
+                    r.swapAt(0, 2);
+                    if (pred(r[1], r[0])) r.swapAt(0, 1);
+                }
+            }
+            else
+            {
+                if (pred(r[1], r[0]))
+                {
+                    r.swapAt(0, 1);
+                }
+                else
+                {
+                    if (pred(r[2], r[1])) r.swapAt(1, 2);
+                }
+            }
+            return;
+        case 4:
+            if (pred(r[1], r[0])) r.swapAt(0, 1);
+            if (pred(r[3], r[2])) r.swapAt(2, 3);
+            if (pred(r[2], r[0])) r.swapAt(0, 2);
+            if (pred(r[3], r[1])) r.swapAt(1, 3);
+            if (pred(r[2], r[1])) r.swapAt(1, 2);
+            return;
+        default:
+            sort5!pred(r[r.length - 5 .. r.length]);
+            if (r.length == 5) return;
+            break;
     }
 
-    immutable maxJ = r.length - 1;
-    for (size_t i = r.length - 2; i != size_t.max; --i)
+    assert(r.length >= 6);
+    /* The last 5 elements of the range are sorted. Proceed with expanding the
+    sorted portion downward. */
+    immutable maxJ = r.length - 2;
+    for (size_t i = r.length - 6; ; --i)
     {
-        size_t j = i;
-
-        static if (hasAssignableElements!Range)
+        static if (is(typeof(() nothrow {
+                auto t = r[0]; if (pred(t, r[0])) r[0] = r[0];
+            }))) // Can we afford to temporarily invalidate the array?
         {
-            auto temp = r[i];
-
-            for (; j < maxJ && pred(r[j + 1], temp); ++j)
+            size_t j = i + 1;
+            if (pred(r[j], r[i]))
             {
-                r[j] = r[j + 1];
+                auto temp = r[i];
+                do
+                {
+                    r[j - 1] = r[j];
```

Comment:
FWIW this is just a memmove and you have random indexing. Wouldn't it be cleaner (and maybe faster) to first search the index of the first `>=` element, and then memmove everything down using an existing algorithm function.